stdlib, python: Add warning message and clarify binary vs metric units

erin-le opened the pull request:
This PR changes memory and cache sizes in various parts of the gem5 codebase to use binary units (e.g. KiB) instead of metric units (e.g. kB). This makes the codebase more consistent, as gem5 automatically converts memory and cache sizes that are in metric units to binary units. 

This PR also adds a warning message to let users know when an auto-conversion from base 10 to base 2 units occurs.

There were a few places in configs and in the comments of various files where I didn't change the metric units, as I couldn't figure out where the parameters with those units were being used.

erin-le committed:
stdlib: Clarify power of 2 vs power of 10

This commit changes files in src/python/gem5 so that memory and
cache sizes use "KiB", "MiB", and "GiB" instead of "KB", etc. This
makes the codebase more consistent, as gem5 will automatically
convert memory and cache sizes that are in metric units (KB) to
binary units (KiB).

Change-Id: If5b1e908ddcff7182b71e789229a3ba1fa6ad1f1

erin-le committed:
python: add warning message for conversion from base 10 to base 2

This commit adds a warning message for when cache or memory sizes
will be automatically converted from metric units (e.g. kB) to
binary units (e.g. KiB).

Change-Id: I4ddf199ff2f00c78bbcb147e04bd88c496fa16ae

BobbyRBruce commented on src/python/m5/util/convert.py:
```diff
@@ -258,6 +264,16 @@ def toMemoryBandwidth(value):
+        and value[-2] in binary_prefixes.keys()
+        and not "i" in value
+    ):
+        print(
+            f"warn: Base 10 memory/cache size {value} will be cast to base 2"
+            + f" size {value[0:-2]}{value[-2].upper()}i{value[-1]}"
+        )
```
Here's a cheap solution to your circular importing. 

You can import anywhere so just do it when execute this function, not when the module is loaded.

```suggestion
        from m5.util import warn
        warn(
            f"Base 10 memory/cache size {value} will be cast to base 2"
            + f" size {value[0:-2]}{value[-2].upper()}i{value[-1]}"
        )
```

erin-le committed:
configs, dev, learning-gem5, python, tests: more clarification

This commit contains the rest of the base 2 vs base 10 cache/memory
size clarifications. It also changes the warning message to use
warn(). With these changes, the warning message should now no
longer show up during a fresh compilation of gem5.

Change-Id: Ia63f841bdf045b76473437f41548fab27dc19631

BobbyRBruce commented on src/python/m5/util/convert.py:
```diff
@@ -258,6 +264,18 @@ def toMemoryBandwidth(value):
 
 
 def toMemorySize(value):
+    if (
+        type(value) is str
+        and len(value) > 1
+        and value[-2] in binary_prefixes.keys()
+        and not "i" in value
+    ):
+        from m5.util import warn
+
+        warn(
+            f"Base 10 memory/cache size {value} will be cast to base 2"
+            + f" size {value[0:-2]}{value[-2].upper()}i{value[-1]}."
+        )
```
I'm not going to blog it on this but i think you could save yourself a lot of effort here by just having a prefix lookup in "src/python/m5/utility/convert.py". Something like:

```py
si_base_10_to_2 = {
    "kB" : "KiB",
    "MB" : "MiB",
     #...  etc.
}
```

then just:

```py
if get_prefix(value) in si_base_10_to_2:
   correct_prefix = si_base_10_to_2
   # ... etc.
```

I also think a function for extractin the units from the string would be handy as demoed in `get_prefix` above.

erin-le committed:
mem, python: refactor error message formatting

This commit refactors the error message added to convert.py.
A mapping between the base 10 and base 2 suffix magnitudes
(e.g. k: ki, M: Mi, etc.) and a new function that extracts the
magnitude and numerical value have been added. Also, a warning
message has been added to the toMemoryBandwidth function in
addition to the one in toMemorySize.

Change-Id: I3ae157d13c7089d38a34a6e4c35a2b58978106d0

BobbyRBruce commented on src/python/m5/util/convert.py:
```diff
@@ -37,6 +37,12 @@
+
+# from m5.util import warn
+# from . import warn
+
```
This comment is no longer needed right? We fixed this?

BobbyRBruce replied:
This still needs addressed.

BobbyRBruce committed:
tests,stdlib,python: Add tests for base 10 to 2 SI unit check

**Note**: Erin needs to complete the commit by expanding this test to
properly test the behavior of this change.

To run the pyunit tests:

```sh
scons build/ALL/gem5.opt -j`nproc`
./build/ALL/gem5.opt tests/run_pyunit.py
```

Change-Id: I8cea0fe8b088e03e84072a000444953768bc3151

BobbyRBruce commented on src/python/m5/util/convert.py:
```diff
@@ -118,6 +133,32 @@ def _split_suffix(value, suffixes):
     return (value[: -len(matches[0])], matches[0]) if matches else (value, "")
 
 
+def _split_suffix_by_prefix(value):
```
I think there should be a way of just having one `_split_suffix` function but i'm getting very confused looking at this files and how it splits these string so am fine with having two functions for now.

BobbyRBruce replied:
In fact, doesn't `_split_suffix(value, suffixes=binary_prefixes)` do this for you?

erin-le replied:
_split_suffix only matches the ending, so there would need to be two additional lists for the B and the B/s endings, e.g. [kB, MB, ...] and [kB/s, MB/s,...]. This seemed worse to me than adding a helper function, but would it actually be preferable?

erin-le replied:
It occurred to me that I could put values through _split_suffix twice instead of using _split_suffix_by_prefix. I think I am going to refactor the code so it does this.

erin-le replied:
It occurred to me that I could put values through _split_suffix twice instead of using _split_suffix_by_prefix. I think I am going to refactor the code so it does this.

BobbyRBruce commented on src/python/m5/util/convert.py:
```diff
@@ -118,6 +133,32 @@ def _split_suffix(value, suffixes):
+    if len(matches) == 2:  # e.g. matches both M and Mi
+        matches.sort(key=len, reverse=True)
+        matches.pop()
+        assert len(matches) <= 1
```
This is a weird assert for two reasons:

1. If `len(matches) == 2` and we pop once then `matches` is always equal to `1`. I can't see anything here that'll ever result in this assert failing.
2. Why `<=` and not just `==`? This login implies there might be a state where `matches == 0` and, in this case, we want to continue in which case the `return` function will fail when it tries to access `matches[0'`.

erin-le replied:
Would it be best just to remove it? It was carried over from _split_suffix, so it looks like it's useful in _split_suffix but not here.

BobbyRBruce replied:
Remove it or change it to `assert len(matches) == 1`. The latter is a pretty meaningless assert but can do no real harm so I'd go with that.

BobbyRBruce reviewed (changes requested):
I've added a PR here demonstrating how to add tests to this: https://github.com/BobbyRBruce/gem5/commit/1ced85887a263966fc78965f198f8eb0a69fe5fe.

My requested changes to this PR is to add tests to ensure this works as intended.

erin-le committed:
python: refactor base 10 to 2 error message

This commit refactors the base 10 to base 2 error message such
that it uses the preexisting _split_suffix function instead
of a new function based off of _split_suffix. This commit also
removes the new helper function used previously.

Change-Id: I44d9ac3d8b98bcff33d6bfea7ffbdb5009272ede

erin-le committed:
tests: add test for checking conversion from base 10 to base 2

This commit adds a test that checks that strings representing
base 10 memory sizes or base 10 memory bandwidths are correctly
converted to strings representing base 2 values.

Change-Id: Ie8cac15f06b4ceb1786484fea4e8ba2111f4e8d3

erin-le committed:
python: remove outdated comment in convert.py

Change-Id: I0cdeb709e5ae1a3100662172d96a5f6328be1a3d

powerjg reviewed (approved):
lgtm. After merging this change, you should probably run grep again and look for any more mistakes in case they sneak in between you creating this change and when it's merged.
